### PR TITLE
add ingress metric

### PR DESCRIPTION
**Pull request by zouyee**
https://github.com/kubernetes/kube-state-metrics/issues/122

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kube-state-metrics/225)

### CODE REVIEW

File: collectors/ingress.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	)
+	descIngressMetadataGeneration = prometheus.NewDesc(
+		"kube_ingress_metadata_generation",
+		"Sequence number representing a specific generation of the desired state.",
+		[]string{"namespace", "name"}, nil,
+	)
+	descIngressLoadBalancer = prometheus.NewDesc(
+		"kube_ingress_loadbalancer",
+		"kube ingress loadbalancer.",
+		[]string{"namespace", "name", "IP", "hostname"}, nil,
```
Comment (andyxning) [inline collectors/ingress.go]: maybe lower case `ip` instead?

File: collectors/ingress.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+func (l IngressLister) List() ([]v1beta1.Ingress, error) {
+	return l()
+}
+
+func RegisterIngressCollector(registry prometheus.Registerer, kubeClient kubernetes.Interface) {
+	client := kubeClient.ExtensionsV1beta1().RESTClient()
+	rslw := cache.NewListWatchFromClient(client, "ingresses", api.NamespaceAll, nil)
+	rsinf := cache.NewSharedInformer(rslw, &v1beta1.Ingress{}, resyncPeriod)
+
+	ingressLister := IngressLister(func() (ingresses []v1beta1.Ingress, err error) {
```
Comment (andyxning) [inline collectors/ingress.go]: `IngressList`?

File: collectors/ingress.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+func (dc *ingressCollector) collectIngress(ch chan<- prometheus.Metric, d v1beta1.Ingress) {
+	addGauge := func(desc *prometheus.Desc, v float64, lv ...string) {
+		lv = append([]string{d.Namespace, d.Name}, lv...)
+		ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, v, lv...)
+	}
+	addGauge(descIngressMetadataGeneration, float64(d.ObjectMeta.Generation))
+	addGauge(descIngressInfo, 1)
+	if d.Status.LoadBalancer.Size() > 0 {
```
Comment (andyxning) [inline collectors/ingress.go]: Not so familiar with `generated.pb.go`. However according to the code for `Size`, it should be ok.

**Comment by zouyee** [comment]
@andyxning PTAL /cc @brancz

File: collectors/ingress_test.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+func TestIngressCollector(t *testing.T) {
+	// Fixed metadata on type and help text. We prepend this to every expected
+	// output so we only have to modify a single place when doing adjustments.
+	const metadata = `
+		# HELP kube_ingress_info The info of ingress.
+		# TYPE kube_ingress_info gauge
+		# HELP kube_ingress_metadata_generation Sequence number representing a specific generation of the desired state.
+		# TYPE kube_ingress_metadata_generation gauge
+	  # HELP kube_ingress_loadbalancer kube ingress loadbalancer.
```
Comment (andyxning) [inline collectors/ingress_test.go]: seems mix using tag and space. @zouyee

File: Documentation/ingress-metrics.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Ingress Metrics
+
+| Metric name| Metric type | Labels/tags |
+| ---------- | ----------- | ----------- |
+| kube_ingress_info | Gauge | `ingress`=&lt;ingress-name&gt; <br> `namespace`=&lt;ingress-namespace&gt;|
+| kube_ingress_metadata_generation | Gauge | `ingress`=&lt;ingress-name&gt; <br> `namespace`=&lt;cronjob-namespace&gt;|
```
Comment (andyxning) [inline Documentation/ingress-metrics.md]: `s/cronjob/ingress`

**Comment by andyxning** [comment]
@zouyee except two nits above, others lgtm.

**Comment by zouyee** [comment]
@andyxning thanks,have done!

**Comment by andyxning** [comment]
/lgtm 

ping @brancz

**Pull request closed without merging**